### PR TITLE
fix: propagate evaluated plans limit and paths limit to the native query planner

### PR DESCRIPTION
**Pull request by goto-bus-stop**
The experimental `max_evaluated_plans` and `paths_limit` options were only propagated to the legacy JS planner, not the native query planner.

Now they work with both planners.

Adds integration tests for the `apollo.router.query_planning.plan.evaluated_plans` metric, with and without `max_evaluated_plans` set, with the legacy and the native query planners.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [x] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[ROUTER-875]: https://apollographql.atlassian.net/browse/ROUTER-875?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apollo-router/src/configuration/mod.rs [hunk 1]
```diff
@@ -416,16 +417,31 @@ impl Configuration {
...
-                apollo_federation::query_plan::query_planner::QueryPlanIncrementalDeliveryConfig {
-                    enable_defer: self.supergraph.defer_support,
-                },
+            incremental_delivery: QueryPlanIncrementalDeliveryConfig {
+                enable_defer: self.supergraph.defer_support,
+            },
             type_conditioned_fetching: self.experimental_type_conditioned_fetching,
-            debug: Default::default(),
+            debug: QueryPlannerDebugConfig {
+                bypass_planner_for_single_subgraph: false,
```
Comment (dariuszkuc) [inline apollo-router/src/configuration/mod.rs:441]: we probably also pass this down ([link](https://github.com/apollographql/router/blob/dev/apollo-federation/src/query_plan/query_planner.rs#L370))?

Comment (goto-bus-stop) [reply apollo-router/src/configuration/mod.rs:441]: there is no router config to enable this option as far as i know. my general feeling is that the option could be removed entirely from the query planner, but 🤷🏻‍♀️

**Comment by goto-bus-stop** [comment]
Wait. Why are there two QP configurations?

https://github.com/apollographql/router/blob/dev/apollo-router/src/query_planner/bridge_query_planner.rs#L169-L182

File: apollo-router/src/configuration/mod.rs (hunk 1 above)
Comment (dariuszkuc) [reply apollo-router/src/configuration/mod.rs:441]: Hmm it is something that we do support in the gateway (which I believe was done as an experimental feature to skip the expensive QP computation) - I'll open a ticket to figure out what we want to do on RS side of things

**Comment by lrlna** [comment]
@goto-bus-stop are you continuing this work still? is it just missing an integration test you wanted to do?

>Wait. Why are there two QP configurations?

Looks like we are not using the `rust_query_planner_config` at all, or at least not in the same way we are using the neighbouring `js_query_planner_config` function. Probably something got lost along the way.

**Comment by goto-bus-stop** [comment]
`rust_query_planner_config` is used for a cache key. so it's important that we have only one source of truth for actual QP & the cache key.

the main thing this needs is a test indeed.

File: apollo-router/src/configuration/mod.rs (hunk 1 above)
Comment (goto-bus-stop) [reply apollo-router/src/configuration/mod.rs:441]: I think it's risky to support this if in the future we move support for authentication directives etc. into the query planner

**Review by lrlna — APPROVED** [review]

File: apollo-router/src/configuration/mod.rs [hunk 2]
```diff
@@ -416,16 +417,31 @@ impl Configuration {
...
+        use apollo_federation::query_plan::query_planner::QueryPlannerConfig;
+        use apollo_federation::query_plan::query_planner::QueryPlannerDebugConfig;
+
+        let max_evaluated_plans = self
+            .supergraph
+            .query_planning
+            .experimental_plans_limit
+            // Fails if experimental_plans_limit is zero; use our default.
+            .and_then(NonZeroU32::new)
+            .unwrap_or(NonZeroU32::new(10_000).expect("it is not zero"));
```
Comment (lrlna) [inline apollo-router/src/configuration/mod.rs:430]: doesn't QueryPlannerDebugConfig already set max_evaluated_plans to `10_000`? 🤔

Comment (goto-bus-stop) [reply apollo-router/src/configuration/mod.rs:430]: Only if you use `QueryPlannerDebugConfig::default()`. The property is required so we would have to do something like this to propagate it correctly:

```rust
let mut debug = QueryPlannerDebugConfig::default();
if let Some(plans_limit) = configuration.supergraph.query_planning.experimental_plans_limit {
    debug.max_evaluated_plans = NonZeroU32::new(plans_limit).unwrap_or(debug.max_evaluated_plans);
}
```
That's even more ceremony and if we use the `Default` impl, and a new option is added to `QueryPlannerDebugConfig`, we could easily forget to pass it on from the router side.

By doing it like this we duplicate the default value but we are at least forced by the compiler to provide all the options intentionally.

Comment (lrlna) [reply apollo-router/src/configuration/mod.rs:430]: makes sense!

**Review by lrlna — APPROVED** [review]

**Pull request merged**